graceful.go

```diff
@@ -1,75 +1,94 @@
 package graceful
 
 import (
+	"context"
 	"os"
 	"os/signal"
-	"sync"
-	"syscall"
 )
 
-// Terminator is the interface that wraps the Terminate method.
-//
-// Terminate instructs the implementing object to shut down,
-// close, disposes, or clean up any associated resources.
-type Terminator interface {
-	Terminate()
+type Manager struct {
+	c   atomicInt64
+	ctx context.Context
+	cf  context.CancelFunc
 }
 
-// TerminatorFunc defines a type on func() as a convenience
-type TerminatorFunc func()
+// New creates a manager with the given context.
+func New() *Manager {
+	ctx, cf := context.WithCancel(context.Background())
+	return &Manager{
+		ctx: ctx,
+		cf:  cf,
+	}
+}
 
-// Terminate implements the Terminator interface on TerminatorFunc.
-func (f TerminatorFunc) Terminate() {
-	f()
+// Go wraps and executes the given function w with a cancellable context.
+func (m *Manager) Go(ctx context.Context, w func(ctx context.Context)) {
+	m.c.inc()
+	go func() {
+		wCtx, cancel := context.WithCancel(ctx)
+		defer func() {
+			cancel()
+			m.c.dec()
+		}()
+
+		go func() {
+			select {
+			case <-m.ctx.Done():
+				cancel()
+			}
+		}()
+		w(wCtx)
+	}()
 }
```

Comment:
Bug: When the function w spawns another go routine with a context, the parent context is terminated.